plugins/nominal-connection-checker/src/type_hierarchy.js

```diff
@@ -309,6 +448,186 @@ export class TypeHierarchy {
+      mappedParams.forEach((param, i) => {
+        // TODO: This doesn't work if they all evaluate to null!
+        if (!param) {
+          // Just ignore nulls, we don't have type info for these params.
+          return;
+        }
+        if (!paramsLists[i]) {
+          paramsLists[i] = [];
+        }
+        paramsLists[i].push(param);
+      });
+    });
+    return paramsLists;
+  }
+
+  /**
+   * Returns an array of arrays of common types for each of the lists of actual
+   * types for a given parameter of the commonDef. Parameter lists are unified
+   * based on the variance of the formal parameter of the common type definition
+   * at that index.
+   * @param {!Array<!Array<TypeStructure>>} paramsLists A list of lists of
+   *     params to unify for a given param of the common def.
+   * @param {!TypeDef} commonDef The common type definition that determines
+   *     the variance of the different parameters.
+   * @param {function(...!TypeStructure):!Array<!TypeStructure>}
+   *     covariantRecursion The function to call in the case of a
+   *     covariant parameter.
+   * @param {function(...!TypeStructure):!Array<!TypeStructure>}
+   *     contravariantRecursion The function to call in the case of a
+   *     contravariant parameter.
+   * @return {!Array<!Array<!TypeStructure>>} An array of arrays of common types
+   *     for each of the lists of actual types for a given parameter of the
+   *     commonDef. An empty subarray means a common type could not be found.
+   */
+  unifyParamsLists(
+      paramsLists,
+      commonDef,
+      covariantRecursion,
+      contravariantRecursion
+  ) {
+    return paramsLists.map((paramList, i) => {
+      switch (commonDef.getParamForIndex(i).variance) {
+        case Variance.CO:
+          return covariantRecursion(...paramList);
+        case Variance.CONTRA:
+          return contravariantRecursion(...paramList);
+        case Variance.INV:
+          // eslint-disable-next-line no-case-declarations
+          const [first, ...rest] = paramList;
+          if (rest.every((typeStruct) => typeStruct.equals(first))) {
+            return [first];
+          }
+          return []; // Empty array means the types do not unify.
+      }
+    });
+  }
+
+  /**
+   * Creates all combinations of elements in the subarrays as arrays. If any
+   * subarray is an empty array, this evaluates to an empty array.
+   * @param {!Array<!Array<*>>} firstArray The first array to add the items
+   *     of the second array onto. Should be an array of arrays for proper
+   *     combinating.
+   * @param {!Array<*>} secondArray An array of elements used to create
+   *     combinations.
+   * @param {!Array<!Array<*>>} rest The rest of the arrays of elements.
+   * @return {!Array<!Array<*>>} All combinations of elements in all of the
+   *     subarrays.
+   * @private
+   */
+  combine_([firstArray, ...[secondArray, ...rest]]) {
```

Comment:
Is this used anywhere but with paramLists?

Reply:
Not at the moment - but in a few PRs it gets moved into a utils file because it starts being used elsewhere.

The reason I separated it out here is b/c I've gotten comments on some past PRs about nested functions being hard to read.

Reply:
Ok. If it wasn't, then I was going to suggest more specific jsDoc for this function.